### ISSUE TITLE
Perplexity of a model

### ISSUE THREAD

**Issue by pwl** [pymc-devs/pymc#2297]
I noticed that in some papers people determine whether the model converged based on [perplexity](https://en.wikipedia.org/wiki/Perplexity#Perplexity_of_a_probability_model) on some data set aside earlier for testing.  Is there a way to compute the perplexity of a model with the new OPVI API?  As far as I understand, to get the perplexity we need to compute the values of `log(q)` of the current variational approximation `q` on user-provided data.

I had a look at the source code and it seems it would amount to evaluating `logq` (or `norm_logq`, I'm not sure at this point) but the complexity of the OPVI API is beyond me at this point and I have no idea how to supply user data to these functions.  Maybe I'm looking in the wrong place and there is an easier way around it.

Long term I hope to implement a perplexity-based convergence test as a callback.

Are the other techniques of evaluating the convergence, apart from simply looking at average loss, available in pymc3?  For example, the Wikipedia article also mentions using cross-entropy as the test function.

**Comment by junpenglao** [comment]
You can find the convergence evaluation callback for VI in PyMC3 here https://github.com/pymc-devs/pymc3/blob/master/pymc3/variational/callbacks.py#L31
So far we are doing it in by looking at the change of loss in a given window. 

As for evaluating the values of `log(q)` maybe @ferrine knows a quick way to do that.

**Comment by pwl** [comment]
@junpenglao Doesn't that callback simply look track a change in the norm of the overall parameter vector of approximations `q_i` (so means and stdevs of transformed probability distributions) instead looking at the loss?  In particular, it is completely ignoring `loss` in its `__call__` definition.

**Comment by ferrine** [comment]
@pwl Yes, We are currently looking at parameters as it works for both parametric gradient and non parametric gradient descent. 
I hope I correctly understood the main idea of the proposed method. Evaluating likelihood for approximate model is quite easy. You can get access to `approx.model.observed_RVs` in callback and in the first iteration compile a function using `approx_lik = approx.apply_replacements(likelihood)`. Note that likelihood is scaled to the full dataset there if using minibatches

**Comment by pwl** [comment]
@ferrine thanks for the tip.  How do I access `likelihood`?  Is it stored somewhere in `approx`?

**Comment by ferrine** [comment]
I think not yet. But you can make it
```python

# first, you create a tensor, that can be any expression that uses pymc3 random nodes
# in our case it is likelihood
likelihood = tt.sum([v.logpt for v in approx.model.observed_RVs])
# to evaluate it under approximate posterior you can replace random nodes with approximation
# for the only sample it is
approx_lik = approx.apply_replacements(likelihood)
# for many samples
approx_liks = approx.sample_node(likelihood, n_samples)
# then you can teke expectation under posterior.
marginal_lik = approx_liks.mean(0)
# after all, function can be compiled
marginal_lik_fn = theano.function([], marginal_lik)
```

**Comment by junpenglao** [comment]
We are currently migrating all the questions to https://discourse.pymc.io
Please continue the discussion [here](https://discourse.pymc.io/t/perplexity-of-a-model/25).
